### PR TITLE
Institute-level soft filters for variants

### LINKED ISSUE #5201: Enable filtering on matched normal
**⚠️ Important:** Please avoid sharing any sensitive or personal information, including text or images, in this issue.

**Is your feature request related to a problem in the current program to new available techology or software? Please describe and add links/citations if appropriate.**

In the next release of balsamic we're planning to enable soft-filtering of SNVs present in the matched normal (the current matched normal filters will be unchanged but will not be hard-filtered), and as a result variants will be uploaded to Scout with a couple of new filters. See issue in MTP-BALSAMIC: https://github.com/Clinical-Genomics/MTP-BALSAMIC/issues/1

At the moment we have 2 filters which are both used to filter out variants based on presence in the normal. There is "germline_risk" from TNscope, and a custom bcftools filter which used to be called "high_normal_tumor_af_frac" but which we will rename in balsamic v17.0.0 to "in_normal". 

Image

This was requested by some customers mainly due to the risk of filtering out clinically relevant variants with a high presence in the normal, but will cause some institutes in Scout to receive variants they are not used to seeing, and which they likely do not wish to see, and to avoid this it would be nice to have easy option to filtering out these "high normal" variants by default for these institutes by applying the balsamic soft-filters.

**Describe the solution you'd like**

We discussed adding another checkbox here:

Image

Which can be set to default removing the soft filtered normal variants on an institute level. But I trust in the scout-devs to figure out the best solution! 🙏 

**Describe alternatives you've considered**

**Additional context**

This feature is planned to be activated on the balsamic side in release 17.0.0, for all T+N matched panel analyses, which will increase the number of variants uploaded to Scout for this analysis type. But as the common germline variants are mostly filtered out by population databases the increase shouldn't be that massive. They will in essence in terms of number of variants become like tumor only analyses, which based on what I have seen would roughly increase the number of variants from 2k to 4k, but it varies on the panel.

To get a sense of what this will mean for the total number of variants uploaded from balsamic I counted the current number of tumor only cases in the production directory to 187, and the number of tumor normal cases to 19. Which means that roughly for every 10 tumor only case there's a tumor normal. As a very rough guesstimate the total number of variants from balsamic should with this feature increase by roughly 5%. 

But together with new filters planned to be added in 17.0.0 probably the net variants will actually decrease. It could also be looked into how this feature could be selectively implemented for certain institutes / or optionally per order.

### PR DESCRIPTION
**Pull request by northwestwitch**
I'm not a big fan of custom-made solutions that add a lot of code but are used only by a group. So in order to fix #5201 I'm proposing this more general feature which can allow filtering on basically any string present in variant["filters"]

This PR adds a functionality or fixes a bug.
- Closes #5201 

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. Go to a case and check number of variants. Example with demo cancer case:

image

2.  As an admin, go to institute's settings and add custom soft filters to filter institute's variants with (can be any string present under the "filters" key in a variant's document. Example:

image

3. Back to variants page, you don't get the same results as before:

image

4. On the filters it tells you which default filters are in place:

image

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by DN
- [x] tests executed by DN

### CODE REVIEW

**Comment by northwestwitch** [comment]
Ready for review now

**Comment by northwestwitch** [comment]
This one needs some refinement, after having tested with a demo case. I'm marking it as in progress

**Comment by northwestwitch** [comment]
Tested on stage using the following filters on cust000:

image


And [this case](https://scout-stage.scilifelab.se/cust000/10-2-TWISTpancan4/cancer/variants?variant_type=clinical):

### Default filters applied:

image

### Include soft filtered vars:

image

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ About changelog [here](https://keepachangelog.com/en/1.0.0/)
 - Enhanced SNV and SV filtering for cancer and rare disease cases, now supporting size thresholds (≥ or < a specified base pair length)
 - Option to exclude ClinVar significance status in SNVs filters form
 - Made HRD a config parameter and display it for cancer cases.
+- Preset institute-level soft filters for variants (filtering based on "filters" values on variant documents). Settings editable by admins on the institute's settings page
```
Comment (dnil) [inline CHANGELOG.md:19]:
```suggestion
- Preset institute-level soft filters for variants (filtering based on "filters" values on variant documents). Settings editable by admins on the institute's settings page. Allows e.g. hiding tumor `in_normal` and `germline_risk` filter status variants.
```

File: scout/adapter/mongo/query.py [hunk 2]
```diff
@@ -373,6 +376,11 @@ def build_query(
 
         return mongo_query
 
+    def soft_filters_query(self, query: dict, mongo_query: dict):
+        """Adds info to variants query to exclude variants flagged by specific filters."""
+        if query.get("institute_soft_filters"):
+            mongo_query["filters"] = {"$nin": query["institute_soft_filters"].split(",")}
```
Comment (dnil) [inline scout/adapter/mongo/query.py:382]: 🥇 This is a very clever little operator!

File: scout/server/blueprints/institutes/templates/overview/institute_settings.html [hunk 3]
```diff
@@ -104,6 +104,13 @@
       placeholder: "Add Sanger email",
     });
 
+    $('#soft_filters').select2({
+      tags: true,
+      theme: 'bootstrap-5',
+      tokenSeparators: [','],
+      placeholder: "tnscope:germline_risk",
```
Comment (dnil) [inline scout/server/blueprints/institutes/templates/overview/institute_settings.html:111]: Wasn't it only 
```suggestion
      placeholder: "germline_risk",
```

**Review by dnil — APPROVED** [review]
Looks good, works well, and much approve of a more general solution that can be adapted later. 

I can't keep from saying I am still far from 100% convinced of the loading 5-10 times more variants in general for solving an edge case (and at the same time providing less than ideal germline analysis, risking that someone decides to skip the full one). It did seem quite a bit more additional variants in the test case, and I have not heard a plan for how to order these separately yet. But that is mostly up to procurement for new db servers for sharing, and someone keeping track that these patients get a full germline workup as well. 😁

File: scout/server/blueprints/institutes/templates/overview/institute_settings.html (hunk 3 above)
Comment (northwestwitch) [reply scout/server/blueprints/institutes/templates/overview/institute_settings.html:111]: Ah yes, that was a previous (now outdated) version. Thanks!
